memory/src/main/scala/filodb.memory/data/OffheapLFSortedIDMap.scala

```diff
@@ -246,22 +286,173 @@ class OffheapLFSortedIDMapReader(memFactory: MemFactory, holderKlass: Class[_ <:
     innerBinSearch(0, mapLen)
   }
 
+  /**
+   * Acquire exclusive access to this map, spinning if necessary. Exclusive lock isn't re-entrant.
+   */
+  def acquireExclusive(inst: MapHolder): Unit = {
+    // Spin-lock implementation. Because the owner of the shared lock might be blocked by this
+    // thread as it waits for an exclusive lock, deadlock is possible. To mitigate this problem,
+    // timeout and retry, allowing shared lock waiters to make progress. The timeout doubles
+    // for each retry, up to a limit, but the retries continue indefinitely.
+
+    var timeoutNanos = 1.millisecond.toNanos
+    var warned = false
+
+    while (true) {
+      if (tryAcquireExclusive(inst, timeoutNanos)) {
+        return
+      }
+
+      timeoutNanos = Math.min(timeoutNanos << 1, 1.second.toNanos) // limit timeout
+
+      if (!warned && timeoutNanos >= 1.second.toNanos) {
+        _logger.warn(s"Waiting for exclusive lock: $inst")
```

Comment:
What is the preferred way to provide such a counter in the code today?

A mitigation strategy to simply release all the shared locks would cause more harm than good. The lock count could end up going negative as share locks are released.

Reply:
We use Kamon to report metrics. One place to put the metric would be inside `object OffheapLFSortedIDMap`. You can see `TimeSeriesShardStats` for some examples.

Reply:
Added some counters.